// test/mocks.js
/**
 * Mocks for the test suite.
 */

function mockDependencies(jest) {
  jest.mock('googleapis');
  jest.mock('remark');
  jest.mock('remark-html');
}

function unmockDependencies(jest) {
  jest.unmock('googleapis');
  jest.unmock('remark');
  jest.unmock('remark-html');
}

// createReadableStream mock return value readableStream
const mockReadableStream = 'readable';
const mockWriteResult = 'written';
const mockTransformResult = 'transformed';
const mockStream = 'mockStream';
const emulateError = new Error('emulateError');
class MockTransform {
  push() {}
  end() {}
};

const mockExtractTransformImport = jest.fn(
  () => Promise.resolve(mockStream)
);

function mockIndex(jest) {
  mockDependencies(jest);
  jest.mock('./lib/extract-transform', () => ({
    __esModule: true,
    default: mockExtractTransformImport,
    iAmAMock: () => {}
  }));
}

function unmockIndex(jest) {
  jest.unmock('./lib/extract-transform');
  unmockDependencies(jest);
}

function mockFs(jest) {
  const mockWriteFile = jest.fn((path, data) => {
    if (data.message === emulateError.message) {
      return Promise.reject(emulateError);
    }
    return Promise.resolve({
      path,
      data
    });
  });

  jest.mock('fs', () => ({
    promises: {
      writeFile: mockWriteFile
    }
  }));

  return mockWriteFile;
}

function unmockFs(jest) {
  jest.unmock('fs');
}

function mockLoad(jest) {
  mockFs(jest);
  jest.mock('stream', () => ({
    Transform: MockTransform
  }));
}

function unmockLoad(jest) {
  jest.unmock('stream');
  unmockFs(jest);
}

const mockTestChunk = 'myspecialtestchunk';

function mockOn(name, cb) {
  if (name ==='data' && this.on.skipError) {
    cb(Buffer.from(mockTestChunk));
  } else if (name === 'end' && this.on.skipError) {
    cb();
  } else if (name === 'error' && !this.on.skipError) {
    cb(emulateError);
  }
  return this;
}

function GoogleAuth() {
}

const defaultMockFiles = [{
  id: '101010',
  name: 'test-file.passthru'
}];

const _mockFiles = [];

const mockGoogleapis = {
  google: {
    auth: {
      GoogleAuth
    },
    drive: () => ({
      files: {
        export: () => Promise.resolve({
          data: {
            on: mockOn
          }
        }),
        list: () => {
          return Promise.resolve({
            data: {
              get files() {
                return _mockFiles.length > 0 ? _mockFiles : defaultMockFiles;
              }
            }
          })
        }
      }
    })
  }
};

function mockExtractTransform(jest) {
  jest.mock('googleapis', () => mockGoogleapis);
  jest.mock('remark-html');
  jest.mock('remark', () => () => ({
    use: () => ({
      process: (inputData, cb) => {
        if (inputData.mockProcessError) {
          cb(inputData.mockProcessError);
        }
        cb(null, inputData.mockProcessResult);
      }
    })
  }));
}

function unmockExtractTransform(jest) {
  jest.unmock('googleapis');
  jest.unmock('remark');
  jest.unmock('remark-html');
}

function mockFiles(files) {
  _mockFiles.length = 0;
  Array.prototype.push.apply(_mockFiles, files);
}

function unmockFiles() {
  _mockFiles.length = 0;
}

module.exports = {
  mockReadableStream,
  mockWriteResult,
  mockTransformResult,
  mockStream,
  mockIndex,
  unmockIndex,
  mockLoad,
  unmockLoad,
  emulateError,
  mockTestChunk,
  mockOn,
  mockGoogleapis,
  mockFs,
  unmockFs,
  mockFiles,
  unmockFiles,
  mockExtractTransform,
  unmockExtractTransform
};
